src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs

```diff
@@ -378,11 +382,77 @@ private async Task EnsureIndexExists(string indexName, ElasticClient esClient)
             }
         }
 
-        private string GetIndexName(ElasticSearchConnectionData connectionData)
+        protected internal string GetIndexName(ElasticSearchConnectionData connectionData)
+        {
+            string dateString = "";
+            try
+            {
+                dateString = ConvertIndexFormat(connectionData.Configuration.IndexFormat);
+            }
+            catch (Exception ex)
+            {
+                string errorMessage = $"{nameof(ElasticSearchOutput)}: indexFormat could not parse as date: {ex.Message}{Environment.NewLine}" +
+                                      $"Invalid IndexFormat value: {connectionData.Configuration.IndexFormat}{Environment.NewLine}" +
+                                      $"Falling back to default value: {ElasticSearchOutputConfiguration.DefaultIndexFormat}";
+                // Changing in connectionData.Configuration so we don't report this every log message
+                connectionData.Configuration.IndexFormat = ElasticSearchOutputConfiguration.DefaultIndexFormat;
+                healthReporter.ReportWarning(errorMessage);
+                dateString = string.Format($"{{0:{connectionData.Configuration.IndexFormat}}}", DateTimeOffset.UtcNow);
+            }
+
+            return connectionData.Configuration.IndexNamePrefix + dateString;
+        }
+
+        public Func<DateTimeOffset> Now { get; set; } = DefaultNow;
+        public static DateTimeOffset DefaultNow() { return DateTimeOffset.UtcNow; }
+
+        protected internal string ConvertIndexFormat(string dateString)
+        {
+            var index = dateString.IndexOf('q');
+            if (index > -1)
+                dateString = dateString.Substring(0, index) + GetQuarterOfYear() + dateString.Substring(index + 1);
+            index = dateString.IndexOf('w');
```

Comment:
Since EventFlow does not support dynamic reconfiguration, you could examine the `dateString` in the constructor of the output and determine what kind of treatment you want to use to get the final `dateString` that you pass to `string.Format`. This would avoid parsing the `dateString` for every batch of events.

 For all special formatting characters that you are introducing here the treatment for determining what the effective `dateString` should be is basically the same:
1. A prefix (does not change from event batch to the next)
2. A string produced by a method (`GetQuarterOfYear` etc) with a single input (the ""now" timestamp) i.e. `Func<DateTimeOffset, String>`
3. A suffix (does not change from event batch to the next)
4. Combine 1 + 2 + 3 to get the effective `dateTime` string.

In fact, even if the user does not use quarter/week of month/week of year, the above generalization still works with prefix=original `dateString`, the method = null, and the suffix = empty string. 

Of course this works if you do not want to support more than one of your special formatting characters ☺️ (e.g. no quarter AND week of year in the same formatting string). But I assume this is the intention here, and we could just document it.

Not sure if all this makes things faster enough to make the additional code complexity worthwhile, so feel free to skip this suggestion.

Reply:
Moving format character lookup to Initialize(), I didn't notice that section before. I am just saving the indexes in local private int fields. Let me know if there is a better way of doing that.

Reply:
I'm not sure if I understand your comments about batches. I don't really know how this all works in the background. Are you suggesting I actually cache indexFormat post-conversion for some period of time? I think moving the character lookups to Initialize() is the best I can do, and is worth the little bit of extra complexity.

I DO want to allow more than one special format character. I personally don't see the point in using more than one, but I don't want to limit others who might have different needs than I do. In fact, I previously limited to using only one of w, W, or WW, but technically, only W and WW are truly incompatible, so I now allow lowercase w and one of the upper case ones.

Reply:
You can disregard my comments about batching etc. I like what you did here; it is simple and effective.